File: app/backend/src/routes/docs.ts
import {Router, Response} from 'express';
import {readFileSync, existsSync} from 'fs';
import path from 'path';
import {config} from '../config';
import {requireAuth} from '../middleware/auth';
import {AuthenticatedRequest, DocTocSection} from '../types';

const router = Router();

function parseTocMarkdown(md: string): { sections: DocTocSection[] } {
    const sections: DocTocSection[] = [];
    let current: DocTocSection | null = null;

    for (const line of md.split('\n')) {
        const sectionMatch = line.match(/^## (.+)/);
        if (sectionMatch) {
            current = {title: sectionMatch[1], items: []};
            sections.push(current);
            continue;
        }
        const itemMatch = line.match(/^- \[(.+?)]\((.+?)(?:#.+?)?\)/);
        if (itemMatch && current) {
            if (!current.items.some(i => i.path === itemMatch[2])) {
                current.items.push({title: itemMatch[1], path: itemMatch[2]});
            }
        }
    }

    return {sections};
}

let tocCache: { sections: DocTocSection[] } | null = null;

router.get('/toc', requireAuth, (_req: AuthenticatedRequest, res: Response) => {
    try {
        if (!tocCache) {
            const md = readFileSync(path.join(config.docsPath, 'toc.md'), 'utf-8');
            tocCache = parseTocMarkdown(md);
        }
        res.json(tocCache);
    } catch {
        res.status(500).json({error: 'Failed to load table of contents'});
    }
});

process.on('SIGHUP', () => {
    tocCache = null;
});

router.get('/content', requireAuth, (req: AuthenticatedRequest, res: Response) => {
    const docPath = req.query.path as string;
    if (!docPath) return res.status(400).json({error: 'Missing path parameter'});

    // Block directory traversal
    const normalized = path.normalize(docPath);
    if (normalized.includes('..') || path.isAbsolute(normalized)) {
        return res.status(400).json({error: 'Invalid path'});
    }

    const full = path.join(config.docsPath, normalized);
    if (!existsSync(full)) return res.status(404).json({error: 'Not found'});

    try {
        res.json({content: readFileSync(full, 'utf-8')});
    } catch {
        res.status(500).json({error: 'Failed to read document'});
    }
});

export default router;

File: app/backend/internal/handlers/docs.go
package handlers

import (
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"

	"github.com/mephalrith/noodles/backend/internal/config"
	"github.com/mephalrith/noodles/backend/internal/errs"
	"github.com/mephalrith/noodles/backend/internal/model"
	"github.com/mephalrith/noodles/backend/internal/respond"
	"github.com/mephalrith/noodles/backend/internal/services"
)

var (
	sectionRE = regexp.MustCompile(`^## (.+)`)
	itemRE    = regexp.MustCompile(`^- \[(.+?)]\((.+?)(?:#.+?)?\)`)
)

type tocCache struct {
	mu       sync.RWMutex
	sections []model.DocTocSection
	loaded   bool
}

var toc tocCache

func parseTocMarkdown(md string) []model.DocTocSection {
	var sections []model.DocTocSection
	var current *model.DocTocSection

	for _, line := range strings.Split(md, "\n") {
		if m := sectionRE.FindStringSubmatch(line); m != nil {
			sections = append(sections, model.DocTocSection{Title: m[1]})
			current = &sections[len(sections)-1]
			continue
		}
		if m := itemRE.FindStringSubmatch(line); m != nil && current != nil {
			duplicate := false
			for _, item := range current.Items {
				if item.Path == m[2] {
					duplicate = true
					break
				}
			}
			if !duplicate {
				current.Items = append(current.Items, model.DocTocItem{Title: m[1], Path: m[2]})
			}
		}
	}

	return sections
}

func ResetTocCache() {
	toc.mu.Lock()
	defer toc.mu.Unlock()
	toc.loaded = false
	toc.sections = nil
}

func HandleDocsToc(cfg *config.Config) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		toc.mu.RLock()
		if toc.loaded {
			sections := toc.sections
			toc.mu.RUnlock()
			respond.OK(w, map[string]any{"sections": sections})
			return
		}
		toc.mu.RUnlock()

		toc.mu.Lock()
		defer toc.mu.Unlock()

		if toc.loaded {
			respond.OK(w, map[string]any{"sections": toc.sections})
			return
		}

		data, err := os.ReadFile(filepath.Join(cfg.DocsPath, "toc.md"))
		if err != nil {
			services.Logger.Error("Failed to load table of contents", "error", err)
			respond.Error(w, errs.Internal("Failed to load table of contents"))
			return
		}

		toc.sections = parseTocMarkdown(string(data))
		toc.loaded = true

		respond.OK(w, map[string]any{"sections": toc.sections})
	}
}

func HandleDocsContent(cfg *config.Config) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		docPath := r.URL.Query().Get("path")
		if docPath == "" {
			respond.Error(w, errs.MissingPath)
			return
		}

		normalized := filepath.Clean(docPath)
		if strings.Contains(normalized, "..") || filepath.IsAbs(normalized) {
			respond.Error(w, errs.InvalidPath)
			return
		}

		full := filepath.Join(cfg.DocsPath, normalized)
		data, err := os.ReadFile(full)
		if err != nil {
			if os.IsNotExist(err) {
				respond.Error(w, errs.NotFound)
				return
			}
			respond.Error(w, errs.Internal("Failed to read document"))
			return
		}

		respond.OK(w, map[string]string{"content": string(data)})
	}
}
